src/crimson/osd/pg_backend.cc

```diff
@@ -766,53 +767,52 @@ PGBackend::write_iertr::future<> PGBackend::writefull(
-  ::crimson::interruptible::interruptible_errorator<
-    ::crimson::osd::IOInterruptCondition,
-    mocked_load_clone_obc_ertr>;
-
-static mocked_lock_clone_obc_iertr::future<crimson::osd::ObjectContextRef>
-mocked_load_clone_obc(const auto& coid)
-{
-  return crimson::ct_error::enoent::make();
-}
-
-static auto head2clone(const hobject_t& hoid)
-{
-  // TODO: transform hoid into coid
-  return hoid;
-}
-
 PGBackend::rollback_iertr::future<> PGBackend::rollback(
-  const SnapSet &ss,
   ObjectState& os,
   const OSDOp& osd_op,
   ceph::os::Transaction& txn,
   osd_op_params_t& osd_op_params,
-  object_stat_sum_t& delta_stats)
+  object_stat_sum_t& delta_stats,
+  crimson::osd::ObjectContextRef head,
+  crimson::osd::ObjectContextLoader& obc_loader)
 {
+  const ceph_osd_op& op = osd_op.op;
+  snapid_t snapid = (uint64_t)op.snap.snapid;
   assert(os.oi.soid.is_head());
   logger().debug("{} deleting {} and rolling back to old snap {}",
-                  __func__, os.oi.soid, osd_op.op.snap.snapid);
-  return mocked_load_clone_obc(
-    head2clone(os.oi.soid)
-  ).safe_then_interruptible([](auto clone_obc) {
-    // TODO: implement me!
-    static_cast<void>(clone_obc);
-    return remove_iertr::now();
-  }, crimson::ct_error::enoent::handle([this, &os, &txn, &delta_stats] {
-    // there's no snapshot here, or there's no object.
-    // if there's no snapshot, we delete the object; otherwise, do nothing.
-    logger().debug("rollback: deleting head on {}"
-                   " because got ENOENT|whiteout on obc lookup",
-                   os.oi.soid);
-    return remove(os, txn, delta_stats, true /*whiteout*/);
-  }), mocked_load_clone_obc_ertr::assert_all{
-    "unexpected error code in rollback"
-  });
+                  __func__, os.oi.soid ,snapid);
+  hobject_t target_coid = os.oi.soid;
+  target_coid.snap = snapid;
+  return obc_loader.with_clone_obc_only<RWState::RWREAD>(
+    head, target_coid,
+    [this, &os, &txn, &delta_stats, &obc_loader]
+    (auto clone_obc) {
+    logger().debug("PGBackend::rollback: loaded clone_obc: {}",
+                   clone_obc->obs.oi.soid);
+    // 1) Delete current head
+    txn.remove(coll->get_cid(), ghobject_t{os.oi.soid,
+                                ghobject_t::NO_GEN, shard});
+    // 2) Clone correct snapshot into head
+    txn.clone(coll->get_cid(), ghobject_t{clone_obc->obs.oi.soid},
+                               ghobject_t{os.oi.soid});
+    //    Copy clone obc.os.oi to os.oi
+    os.oi.clear_flag(object_info_t::FLAG_WHITEOUT);
+    os.oi.copy_user_bits(clone_obc->obs.oi);
+    // TODO: 3) Calculate clone_overlaps by following overlaps
+    //          forward from rollback snapshot
+    return rollback_iertr::now();
+  }).handle_error_interruptible(
+    crimson::ct_error::enoent::handle([this, &os, &snapid] {
+      logger().error("PGBackend::rollback: {} clone object"
+                     " with snap_id of {} not found",
+                     os.oi.soid, snapid);
```

Comment:
Isn't this case supposed to be a delete?

Reply:
Please see c5218903239564dea272b63ab4ec2e0fc696836a.
I couldn't use `remove()` with `handle_error_interruptible` constrains.
One way to solve that was using the `delete_head` flag, do you find that acceptable?